v2/lint/source.go

```diff
@@ -25,17 +25,18 @@ import (
 type LintSource string
 
 const (
-	UnknownLintSource        LintSource = "Unknown"
-	RFC5280                  LintSource = "RFC5280"
-	RFC5480                  LintSource = "RFC5480"
-	RFC5891                  LintSource = "RFC5891"
-	CABFBaselineRequirements LintSource = "CABF_BR"
-	CABFEVGuidelines         LintSource = "CABF_EV"
-	MozillaRootStorePolicy   LintSource = "Mozilla"
-	AppleCTPolicy            LintSource = "Apple"
-	ZLint                    LintSource = "ZLint"
-	AWSLabs                  LintSource = "AWSLabs"
-	EtsiEsi                  LintSource = "ETSI_ESI"
+	UnknownLintSource                   LintSource = "Unknown"
+	RFC5280                             LintSource = "RFC5280"
+	RFC5480                             LintSource = "RFC5480"
+	RFC5891                             LintSource = "RFC5891"
+	CABFBaselineRequirementsOverbearing LintSource = "CABF_BR_Overbearing"
```

Comment:
Howdy @RufusJWB, I'll be taking over a good deal of the day-to-day maintenance of the repo so I'll be getting us over the finish line with these PRs. 

I have briefly reviewed the pull request that preceded this one, and it seems that I'm about to continue the thread on this particular issue with more windy roads :smile:

It seems as though this category of linter was added because a certificate could be an OCSP Delegated Responder and _not_ have a `serverAuth` EKU. Normally then, the CAB BR would not have an opinion on it. However, the CAB BR _does_ have an opinion on it if the issuing certificate (or any certificate further up the chain?) _does_ have the `serverAuth` EKU. Is that a correct restatment?

However, since we do not have a copy of the certificate chain, we cannot verify whether or not the CAB BR applies, so we simply call it a warning and put it in the category of "Overbearing" (which I reckon is hinting at the notion that the category is the CAB BR possibly "overstepping" its bounds).

But what if we _did_ have a copy of the certificate chain? Would that allow us to disambiguate this?

```go
// Nothing in the chain is a TLS cert, so warn on CAB.
func (w* WarningLint) CheckApplies(c *x509.Certificate) bool {
	return util.IsDelegatedOCSPResponderCert(c) && !util.Any(c.Chain, util.IsServerAuthCert)
}
// Something in the chain is a TLS cert, so error on CAB.
func (w* ErrorLint) CheckApplies(c *x509.Certificate) bool {
	return util.IsDelegatedOCSPResponderCert(c) && util.Any(c.Chain, util.IsServerAuthCert)
}
```
Note that I'm not 100% on the semantics of that example, but it was just an example.

@zakird The easiest way I can currently think to accomplish something like this is to update zcrypto's `Certificate` type to actually retain the chain so that linters can access it if present (with minimal changes to ZLint aside from the obvious CLI update that would have to occur). I reckon that these sorts of hacks are the exact reason why the fork exists, right? I'd be happy to open a PR if we like this idea.

Reply:
Well, one of the big issues is that there are many chains, and the violation might depend on the context, or the result might change over time (as new chains become available).

Today, things are hermetic: the results for a cert are consistent over time (at least, for the same version). Adding the chain makes it a bit complex.

It seems that there’s room for offering an API here, but it would require fleshing out a bit of a proposal. There’s a big question on whether it’s better to:
  1. Supply a single chain, and perform multiple invocations
  2. Provide multiple chains in one call.

There are good arguments for both, such as performance and usability, so it’s probably worth considering on a new issue and noodling about the design space here.

Reply:
This is discussed here: https://github.com/zmap/zlint/issues/428. As @sleevi mentions, it's a bit unfortunate because lint results are now dependent on a (leaf, intermediate) instead of just the (leaf,). But, if it allows us to tests additional functionality, I'm of the opinion: so be it, particularly given that CT already provides the chain in the API, so both monitors and CAs have a pretty good idea what they "should" test. We need to change the API though to accept parent certificates optionally, and likely a new output status.

Reply:
Hi Christopher! Thank you for taking over here and I'm very interested to get this PR through the door. It was a long trip so far. I would love to implement your proposal, but one problem I'm seeing with it is, that it collides with https://github.com/zmap/zlint/issues/476 where it was stated, that one lint should have only one result, either `WARN`or `ERROR`. So basically we would need to move the `util.Any(c.Chain, util.IsServerAuthCert)` into the `isApplicabable` method of both lints. What do you think?

Reply:
So, I still like the idea of extending the CLI to accept zero-or-more certificate chains as I reckon that this is not the last time that a linter is going to want to have the chain around in order to do its business. However, that is a much larger architectural change that Id' rather do on my own time instead of blocking this PR on it (as Zakir and Sleevi point out there are some interesting implications and considerations to this).

Besides, I believe that the core issue would still be there even if we did implement such a system. That is, if an operator runs this lint without providing at least one certificate chain, then what should we output here? A `Warning` that says in the comments that this might actually be an `Error`? People are used to warnings complaining about `MAY` clauses whereas this particular warning is actually saying that they _might_ be in violation of a `SHALL`. That sounds like a sneaky pitfall.

How about a new `lint.LintStatus` ,`Unproven`? This can stand in places where we want to say, "listen, this lint _might_ apply or it _might_ pass but we don't have enough information to know for sure". It's a notch down from a failure, but is more eyebrow raising than a warning.

Reply:
It seems to me that there are just two lint conditions, one is only relevant if the parent certificate chain was provided. In our systems, we have had constant questions about how to perform lints where the certificate chain is relevant and, since that currently isn't within the purview of zlint, we have had to manually implement those lints. I'm not sure we can simply ignore the benefits of adding the chain to the set of lints that could then be created.

That said, I still agree with the concern I raised on #428 that it seems particularly heavy to modify the Lint interface for every Lint when only a subset of Lints requires an issuing chain to be passed in. It will also cause incompatibilities with external library integrations, so the cost of modifying the Lint interface should be weighed heavily before making such a change. I don't have a good solution to present for this, unfortunately.

Historically, we have preferred to have the lints following the appropriate MAY, SHOULD, and MUST terminology that is present in the guidelines, however in #492 I indicated another area that presents a parallel, but similar, flaw in that translation. The issue is that zlint can't possibly always know enough information to determine if the lint currently being run violates the Status level that the lint was created for. I'm not sure that adding another Status accurately reflects the amount of concern that is warranted via failure in those cases.

Reply:
@cardonator could it be an issue of granularity? That is, completely flesh out the enum to have...

```go
UnprovenWarning = 8
UnprovenFatal = 9
...
```

...so that we can accurately convey that we might have found a `MUST` violation but we couldn't quite prove it. A question remaining there is, "should an `UnprovenFatal` result in a non-zero exit code?".

Reply:
@christopher-henderson hmm, maybe. I think the challenge with a concept like "Unproven" is that it's unclear how concerned you should be (which you correctly pointed out). Currently, zlint failure levels are pretty straightforward, i.e. if you get a Fail it's a big concern, if you get a Warning it's a medium concern and if you get a Notice it's just something interesting to take note of or verify your compliance with.

In the context of this suggestion, what would you change about the ultimate solution to #492? I think that's a good sample for how something like this would be handled, although this particular case is likely more perilous than that one was if research determined a CA was out of compliance.

Reply:
Should we move this discussion to #428 just to capture it all? It sounds like there are elements of the design space being explored, and it might make sense to make sure it's kept in one place (or some of the proposals here written up). Not a big ask, just keeps drawing my attention to this issue, which is a bit different :)

Reply:
Ported this over to #428 and fired up #522 to discuss the issue of ambiguous lint results.

@RufusJWB seeing as we got the other review in for certs that are definitively for TLS, would it be okay for this one to sit on the back-burner until we figure out the above (it just seems less pressing than explicitly TLS certs)?

Reply:
@christopher-henderson I'm sorry, that I come back to this topic only now. They the holiday season is supposed to be the quietest time in the year but in reality, I always have the feeling it is the most stressful one at work... Anyhow, yes I'm fine with waiting for the results of those two discussions.